--- a/api/baseapi.cpp	2014-05-20 13:54:52.208220043 -0400
+++ b/api/baseapi.cpp	2014-05-20 13:54:48.784257349 -0400
@@ -62,7 +62,7 @@
 #include <windows.h>
 #include <stdlib.h>
 #else
-#include <glob.h>
+#include <dirent.h>
 #include <libgen.h>
 #include <string.h>
 #endif
@@ -308,18 +308,29 @@
       FindClose(handle);
     }
 #else
-    glob_t pglob;
-    char **paths;
-    char *path, *dot;
-    if (glob(pattern.string(), 0, NULL, &pglob) == 0) {
-      for (paths = pglob.gl_pathv; *paths != NULL; paths++) {
-        path = basename(*paths);
-        if ((dot = strchr(path, '.'))) {
-          *dot = '\0';
-          langs->push_back(STRING(path));
+    DIR *dir;
+    struct dirent *dirent;
+    char *dot;
+
+    STRING extension = STRING(".") + kTrainedDataSuffix;
+
+    dir = opendir(tesseract_->datadir.string());
+    if (dir != NULL) {
+        while ((dirent = readdir(dir))) {
+            // Skip '.', '..', and hidden files
+            if (dirent->d_name[0] != '.') {
+                if (strstr(dirent->d_name, extension.string()) != NULL) {
+                    dot = strrchr(dirent->d_name, '.');
+                    // This ensures that .traineddata is at the end of the file name
+                    if (strncmp(dot, extension.string(),
+                                strlen(extension.string())) == 0) {
+                        *dot = '\0';
+                        langs->push_back(STRING(dirent->d_name));
+                    }
+                }
+            }
         }
-      }
-      globfree(&pglob);
+        closedir(dir);
     }
 #endif
   }
